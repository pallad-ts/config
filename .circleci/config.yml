version: 2.1

orbs:
  node: circleci/node@4.7

jobs:
  main_job:
    executor:
      name: node/default
      tag: "14.18"
    steps:
      - checkout
      - run: "npm ci"
      - run:
          command: ./node_modules/.bin/lerna run compile
          name: compilation
      - run:
          command: ./node_modules/.bin/lerna run lint
          name: linter
      - run:
          command: ./node_modules/.bin/lerna run test
          name: test
      - store_test_results:
          path: ./test_results

workflows:
  main:
    jobs:
      - main_job
