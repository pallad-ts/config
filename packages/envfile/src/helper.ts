import {EnvFileLoader} from "./EnvFileLoader";
import {DefaultValueDependency, TransformableDependency} from "@pallad/config";
import {EnvFileDependency} from "./EnvFileDependency";

export function createEnvFileHelper(options: EnvFileHelper.Options) {
    const loader = new EnvFileLoader(Array.isArray(options.paths) ? options.paths : [options.paths], options.config);

    return function <T>(key: string, opts?: EnvFileHelper.HelperOptions<T>) {
        return DefaultValueDependency.create(
            new EnvFileDependency(key, opts && opts.transformer, loader),
            opts && opts.defaultValue
        );
    }
}

export namespace EnvFileHelper {
    export interface Options {
        paths: string[] | string,
        config?: Partial<EnvFileLoader.Config>
    }

    export interface HelperOptions<T> {
        transformer?: TransformableDependency.Transformer<T>,
        defaultValue?: T
    }
}