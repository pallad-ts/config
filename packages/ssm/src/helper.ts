import {SSM} from 'aws-sdk';
import {DefaultValueDependency, TransformableDependency, env, FirstAvailableDependency, Dependency} from "@pallad/config";
import {SSMDependency} from "./SSMDependency";
import DataLoader = require("dataloader");

const splitArray = require('split-array');

export function parameterDeserializer(parameter: SSM.Parameter) {
    if (parameter.Type === 'StringList') {
        return parameter.Value!.split(',');
    }
    return parameter.Value;
}

function createDateLoader(batchFN: SSMHelper.BatchFunction) {
    return new DataLoader<string, SSMDependency.Value>(batchFN);
}

export function createSSMHelper(options?: SSMHelper.Options) {
    const ssm = (options && options.ssm) || new SSM();
    const paramDeserializer = (options && options.parameterDeserializer) || parameterDeserializer;

    const batchFn: SSMHelper.BatchFunction = async (keys: readonly string[]) => {
        const map = new Map<string, Exclude<SSMDependency.Value, undefined>>();
        const batches = splitArray(keys, 10);
        for (const batch of batches) {
            const parameters = await ssm.getParameters({
                Names: batch,
                WithDecryption: true
            }).promise();

            if (parameters.Parameters) {
                for (const parameter of parameters.Parameters) {
                    map.set(parameter.Name!, paramDeserializer(parameter));
                }
            }
        }
        return keys.map(x => map.get(x));
    };

    const dataLoader = ((options && options.createDataLoader) || createDateLoader)(batchFn);
    return function <T>(key: string, opts?: SSMHelper.HelperOptions<T>) {
        const finalKey = ((options && options.prefix) || '') + key;
        return DefaultValueDependency.create(
            new SSMDependency(finalKey, opts && opts.transformer, dataLoader),
            opts && opts.defaultValue
        );
    };
}

export const ssm = createSSMHelper();

export namespace SSMHelper {
    export interface Options {
        ssm?: SSM;
        createDataLoader?: (batchFn: BatchFunction) => DataLoader<string, SSMDependency.Value>;
        parameterDeserializer?: ParameterDeserializer;
        prefix?: string;
    }

    export type EnvOrOptions = Options & {
        envHelper?: <T>(key: string, opts?: HelperOptions<T>) => Dependency<T>;
    }

    export type BatchFunction = (key: readonly string[]) => Promise<SSMDependency.Value[]>;

    export type ParameterDeserializer = (parameter: SSM.Parameter) => Promise<any> | any;

    export interface HelperOptions<T> {
        transformer?: TransformableDependency.Transformer<T>,
        defaultValue?: T
    }

    export interface ENVOrSSMHelperOptions<T> extends HelperOptions<T> {
        ssmName?: string;
    }
}

export function createEnvOrSSMHelper(options?: SSMHelper.EnvOrOptions) {
    const ssm = createSSMHelper(options);
    const envHelper = (options && options.envHelper) || env;
    return function <T>(key: string, opts?: SSMHelper.ENVOrSSMHelperOptions<T>) {
        const ssmName = (opts && opts.ssmName) || key;

        const transformer = opts && opts.transformer;

        return DefaultValueDependency.create(
            new FirstAvailableDependency(
                envHelper(key, {transformer}),
                ssm(ssmName, {transformer})
            ),
            opts && opts.defaultValue
        );
    }
}