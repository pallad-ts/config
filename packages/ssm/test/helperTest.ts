import {createEnvOrSSMHelper, createSSMHelper, SSMHelper} from "@src/helper";
import {createEnvHelper, type} from '@pallad/config';
import * as sinon from 'sinon';
import DataLoader = require("dataloader");

describe('helper', () => {
    describe('ssm', () => {
        let helper: ReturnType<typeof createSSMHelper>;

        const CONFIG = {
            prefix: 'pallad-config-'
        };

        beforeEach(() => {
            helper = createSSMHelper({
                ...CONFIG
            });
        });

        describe('loading', () => {
            it('success', () => {
                const dependency = helper('var1');

                return expect(dependency.getValue())
                    .resolves
                    .toEqual('bar')
            });


            it('success with default value', () => {
                const dependency = helper('var100', {defaultValue: 'bar'});
                return expect(dependency.getValue())
                    .resolves
                    .toEqual('bar');
            });

            it('failure due to lack of parameter', () => {
                const dependency = helper('var1000');

                return expect(dependency.getValue())
                    .rejects
                    .toMatchSnapshot();
            });

            it('loading with type conversion', () => {
                const dependency = helper('var4', {transformer: type.int});

                return expect(dependency.getValue())
                    .resolves
                    .toEqual(10);
            });

            it('loading array of strings', () => {
                const dependency = helper('var2');
                return expect(dependency.getValue())
                    .resolves
                    .toEqual(['foo', 'bar']);
            });

            it('uses custom dataLoader factory', () => {

                const dataLoader = {};
                const stub = sinon.stub()
                    .resolves(dataLoader);

                createSSMHelper({
                    ...CONFIG,
                    createDataLoader: stub,
                });

                sinon.assert.calledOnce(stub);
            });

            it('using parameter deserializer', () => {
                const stub = sinon.stub()
                    .resolves(1000);

                const helper = createSSMHelper({
                    ...CONFIG,
                    parameterDeserializer: stub,
                });

                const dependency = helper('var1');

                return expect(dependency.getValue())
                    .resolves
                    .toEqual(1000);
            });

            it('loading many variables', () => {
                const dependencies = [
                    helper('var1'),
                    helper('var2'),
                    helper('var3'),
                    helper('var4'),
                    helper('var5'),
                    helper('var6'),
                    helper('var7'),
                    helper('var8'),
                    helper('var9'),
                    helper('var10'),
                    helper('var11'),
                    helper('var12'),
                ];

                return expect(
                    Promise.all(dependencies.map(x => x.getValue()))
                )
                    .resolves
                    .toEqual([
                        'bar',
                        ['foo', 'bar'],
                        'world',
                        '10',
                        'wjeNdAbmyjtqAU2OcuPR15F4xDzlV1uZHmZNNl1p',
                        "u6yjSXzp9q5yhaCdEWyj",
                        "7FZxN9NHXcgleeP0TSva",
                        "R3aogUWGdvDJy03brMVEZBnqKJYZcp",
                        "XUlAtGjdCwIiHDnojsFqmOC6HO8SyXX",
                        "u97tvnLBUOurP3FcpUWRqbO81GqqH",
                        "HMZNkhmMOxZsylqOaxfNwjeAf9Ju",
                        "DRJeSQn6CmLVuo5eXHGkVABTsgy",
                    ]);
            })
        });
    });

    describe('envOrSSM', () => {
        let helper: ReturnType<typeof createEnvOrSSMHelper>;

        const ENVS = {
            var1: 'foo',
        };
        const CONFIG: SSMHelper.EnvOrOptions = {
            prefix: 'pallad-config-',
            envHelper: createEnvHelper(ENVS)
        };

        beforeEach(() => {
            helper = createEnvOrSSMHelper({
                ...CONFIG
            });
        });

        it('takes env variable since it is defined', () => {
            const dependency = helper('var1');

            return expect(dependency.getValue())
                .resolves
                .toEqual('foo');
        });

        it('takes ssm value since env is not defined', () => {
            const dependency = helper('var2');

            return expect(dependency.getValue())
                .resolves
                .toEqual(['foo', 'bar']);
        });

        it('takes ssm value under different name since env is not defined', () => {
            const dependency = helper('var2', {
                ssmName: 'var3'
            });
            return expect(dependency.getValue())
                .resolves
                .toEqual('world');
        });
    });
});