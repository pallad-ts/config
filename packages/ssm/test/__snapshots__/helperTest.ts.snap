// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper ssm loading failure due to lack of parameter 1`] = `[Error: Config value not available: SSM (pallad-config-var1000)]`;
